### DISCUSSION TITLE
How to completely prohibit caching of the administrative part of the site?

### DISCUSSION THREAD

**Discussion by Spidometrs** [spatie/laravel-responsecache#479] [Q&A]
How to completely prohibit caching of the administrative part of the site?

My site on Laravel 10 has an administrative part https://mysite.com/admin
The administrative part of the site has links of the type:
https://mysite.com/admin/system/users/profile/1
https://mysite.com/admin/pages
https://mysite.com/admin/pages/edit/22

In which directory should I create a file and what code to write to this file in order to completely prohibit the pages of the administrative part of the site?
I ask for help.

**Comment by vinhkhangphung** [comment]
You can use the exposed middleware CacheResponse and DoNotCacheResponse per route or route group to allow/prevent the caching behaviors, as in https://github.com/spatie/laravel-responsecache?tab=readme-ov-file#preventing-a-request-from-being-cached
Installlation guide: https://github.com/spatie/laravel-responsecache?tab=readme-ov-file#installation

**Comment by Spidometrs** [comment]
Laravel 10.

 I added to the app/Http/Kernel.php file:
```
...

protected $middlewareAliases = [
   ...
   'doNotCacheResponse' => \Spatie\ResponseCache\Middlewares\DoNotCacheResponse::class,
   'cacheResponse' => \Spatie\ResponseCache\Middlewares\CacheResponse::class,
];
```
Next, I added to the /routes/web.php file:

```
<?php

/*
|--------------------------------------------------------------------------
| Web Routes
|--------------------------------------------------------------------------
|
| Here is where you can register web routes for your application. These
| routes are loaded by the RouteServiceProvider and all of them will
| be assigned to the "web" middleware group. Make something great!
|
*/

Route::get('/', 'PublicController@getIndex')->middleware('cacheResponse:300');

?>
```
I expected to get a caching of the front -end of my site, but nothing happens.

Next, I changed the record to caching one content page, but received a 500 error.
`Route::get('/contact', 'PublicController@getIndex')->middleware('cacheResponse:300');`
What am I doing wrong?

**Reply by vinhkhangphung** [reply]
I cannot troubleshooting without further info, so may I ask about your current response cache driver? ("array", "database", "file", "memcached", "redis", "dynamodb", "octane", "null")?

**Comment by Spidometrs** [comment]
"redis"

**Reply by khangphung-vl** [reply]
- Have you start Redis yet? Either by keeping terminal alive to run Redis, run as Windows service, Docker, Laragon,... your choice. Test it out by client-cli then type PING, you should receive PONG as result.
- Does it run on the correct port? (usually it's 6379, look at the database.php config file)
- Do you have a `RESPONSE_CACHE_DRIVER=redis` in the env file? It will be used by the **responsecache.php** config file
`'cache_store' => env('RESPONSE_CACHE_DRIVER', 'file'),`
- Have you install any Redis Client for PHP? (either predis or phpredis, I personally use predis) and specify it as `REDIS_CLIENT=predis` in the env file? It will be used by database.php
`'client' => env('REDIS_CLIENT', 'phpredis')`
 (I'm not sure about the the last one, you must try it out yourself)

**Reply by Spidometrs** [reply]
> * Have you start Redis yet? Either by keeping terminal alive to run Redis, run as Windows service, Docker, Laragon,... your choice. Test it out by client-cli then type PING, you should receive PONG as result.
> 
>     * Does it run on the correct port? (usually it's 6379, look at the database.php config file)
> 
>     * Do you have a `RESPONSE_CACHE_DRIVER=redis` in the env file? It will be used by the **responsecache.php** config file
>       `'cache_store' => env('RESPONSE_CACHE_DRIVER', 'file'),`
> 
>     * Have you install any Redis Client for PHP? (either predis or phpredis, I personally use predis) and specify it as `REDIS_CLIENT=predis` in the env file? It will be used by database.php
>       `'client' => env('REDIS_CLIENT', 'phpredis')`
>       (I'm not sure about the the last one, you must try it out yourself)

...

Of course, Redis has already been installed and works great with other scripts.
laravel-responsecache - does not work if I connect it to the Protected $middlewareAliases:
```
...
protected $middlewareAliases = [
   ...
   'doNotCacheResponse' => \Spatie\ResponseCache\Middlewares\DoNotCacheResponse::class,
   'cacheResponse' => \Spatie\ResponseCache\Middlewares\CacheResponse::class,
];
```

If I connect laravel-responsecache to:
```
...

protected $middlewareGroups = [
   'web' => [
       ...
       \Spatie\ResponseCache\Middlewares\CacheResponse::class,
   ],

...

protected $middlewareAliases = [
   ...
   'doNotCacheResponse' => \Spatie\ResponseCache\Middlewares\DoNotCacheResponse::class,
];
```
 then he caching the entire site.
But I need to caching some controllers.
How can I cach some controllers?

**Reply by vinhkhangphung** [reply]
```
protected $middlewareGroups = [
   'web' => [
       ...
       \Spatie\ResponseCache\Middlewares\CacheResponse::class,
   ],
```

This code essentially make the whole `web` route going through the CacheResponse middleware, so as you said, this setting caches everything. I guess Redis is not the issue here. Can you reproduce the problem, if you can, please share it.

**Comment by Spidometrs** [comment]
It is possible that my file /routes/web.php is not complete:

```
<?php

/*
|--------------------------------------------------------------------------
| Web Routes
|--------------------------------------------------------------------------
|
| Here is where you can register web routes for your application. These
| routes are loaded by the RouteServiceProvider and all of them will
| be assigned to the "web" middleware group. Make something great!
|
*/

Route::get('/', 'PublicController@getIndex')->middleware('cacheResponse:300');

?>
```
I still have to add to the code of this file?

**Reply by vinhkhangphung** [reply]
It's either
1. Register the whole `web` route to go through the cacheResponse middleware
```
protected $middlewareGroups = [
   'web' => [
       ...
       \Spatie\ResponseCache\Middlewares\CacheResponse::class,
   ],
```
2. Register each route or route group to go through the cacheResponse middleware
`Route::get('/', 'PublicController@getIndex')->middleware('cacheResponse:300');`

**Reply by Spidometrs** [reply]
> It's either
> 
>     1. Register the whole `web` route to go through the cacheResponse middleware
> 
> 
> ```
> protected $middlewareGroups = [
>    'web' => [
>        ...
>        \Spatie\ResponseCache\Middlewares\CacheResponse::class,
>    ],
> ```
> 
>     2. Register each route or route group to go through the cacheResponse middleware
>        `Route::get('/', 'PublicController@getIndex')->middleware('cacheResponse:300');`

In this case, the entire site is caching.
In this case, the routing does not work for the specified controller.
This option is not suitable for use.
